doc/dev/osd_internals/multi_object_write_transaction.rst

```diff
@@ -0,0 +1,117 @@
+
+     };
+
+     based on this, client could send a group of ops on multi-objects 
+     through *MOSDOp* to the primary osd in charge of the *master* object 
+     which is explicitly specified by client, call this osd *MASTER*
+
+  #. *MASTER* receives the multi-object write operation request, performs 
+     dead lock check (detailed below), extracts *ObjectWriteOperations* 
+     by objects from *slaves*, and sends them to corresponding primary 
+     osds individually, calls those osds *SLAVE*, the transaction 
+     enters the *PREPARE* state
+
+  #. *SLAVE* receives *MOSDOp* from *MASTER*, note each op received by *SLAVE*
+     is a group of write operations on a single slave object. *SLAVE* 
+     performs dead lock check, does the normal *prepare_transaction()* stuff, 
+     if everything is ok, *SLAVE* will lock the object, and delay further 
+     read/write operations on this object until the transaction finishs. 
+     *SLAVE* sends a *QUERY* message to *MASTER*, and returns *MASTER* 
+     *EINPROGRESS*. The *QUERY* message is used for *MASTER* to notify *SLAVE* 
+     when to release transaction context, *MASTER* will not ack *QUERY* until 
+     itself has released transaction context
+
+  #. *MASTER* collects response from *SLAVE*, if all *EINPROGRESS*, then sends
+     *SLAVE* *LOG* message, the transaction enters the *LOG* state
+     
+  #. *SLAVE* receives *LOG*, constructs a write op which persists the 
+     multi-object operation information in the object attribute named 
+     *MOWO_ATTR*, and also synchronizes the op within its replica osds, 
+     utilizing the present single object write transaction mechanism, then 
+     returns *MASTER* zero
+     
+  #. *MASTER* collects response from *SLAVE*, if all zero, persists the 
+     multi-object operation information itself, then performs the write op on 
+     the master object, and synchronizes the op within its replica osds, return 
+     client *ACK*
+     
+  #. *MASTER* sends *SLAVE* *COMMIT & UNLOCK* message, the transaction enters 
+     the *COMMIT* state
+     
+  #. *SLAVE* receives *COMMIT & UNLOCK* message, first processes *COMMIT*, 
+     performs the write op on slave object, and synchronizes the op within its
+     replica osds, returns *MASTER* zero, *SLAVE* processes *UNLOCK*, cleans up 
+     the transaction and deletes the persisted object attribute
+     
+  #. *MASTER* collects response from *SLAVE*, if all zero, clean up the 
+     transaction and deletes the persisted object attribute, acks *SLAVE* 
+     *QUERY* message
+     
+     
+Fault Tolerance
+---------------
+
+- If *SLAVE* down at any time before persisting transaction to object 
+  attribute, when *MASTER* resends *PREPARE* or (re)sends *LOG*, since *SLAVE* 
+  has lost the in-memory transaction information, it will return *EPERM*. If 
+  *MASTER* receives *EPERM*, it recognizes the situation and redo *PREPARE* for 
+  this *SLAVE* to resend the operation to it to proceed
```

Comment:
The master needs to know to do this.  Is the query message system part of the Objecter where this would happen automatically, or is it a special message sent out-of-band from the Objecter?  I think the best way to handle this would be to structure the PREPARE/LOG portion as a linger op hidden behind the Objecter.  Otherwise, we risk having to reimplement a significant portion of the objecter's interval and map tracking stuff.  Also, hiding it behind the objecter also makes it a lot more trivial to deal with different pools.